Add Volumes and Env Variables to container details

tomasmatus opened the pull request:
Following conversation in #923 this adds new tab in container details
which shows ports, mounted volumes and filtered environment variables.

I decided someone might find it useful to check all env variables so I've added `Show more` button which shows all variables.

---
## Display ports, volumes and environment variables

Published ports, mounted volumes and environment variables are now visible in container integration tab.
By default only variables set by user are shown and after clicking `Show more` all variables are visible.

![containers-integration](https://user-images.githubusercontent.com/10246/179953332-a7686942-624c-4c65-970d-cedce0847958.png)

tomasmatus commented:
Apparently I can't request review so I'll do it with a friendly ping instead :) @jelly @garrett

garrett commented:
The layout feels weird to me; I didn't even notice that there were volumes at a first glance.

Either we need to reconsider the layout for horiontal groups a bit (spacing using different widths, for example) or perhaps just make it a vertical list of groups. Something's currently off.

tomasmatus commented:
> The layout feels weird to me; I didn't even notice that there were volumes at a first glance.

I'm sad to hear that. I followed the design used in `Details` tab. I'll try experimenting and see if I can come up with something better.

tomasmatus commented:
This is what it looks like with `spaceItemsXl`
![image](https://user-images.githubusercontent.com/74668142/174755406-1547ae35-ca18-412f-a42c-8e6e9402947b.png)

I don't think making it vertical is a good choice. We'd be wasting horizontal space and with containers like toolbox the list would be very tall.

jelly commented:
The show less/more button works for all container rows which is kinda funny but not what we want :-)

![image](https://user-images.githubusercontent.com/67428/174826906-975380d8-e463-44d7-b610-7a6ec53ceefb.png)

jelly commented on src/ContainerIntegration.jsx:
```diff
@@ -0,0 +1,120 @@
+        <ListItem key='show-more-env-button'>
+            <Button variant='link' isInline
+                onClick={() => onShowMoreEnvVariablesClick()}>
+                {showMore ? 'Show less' : 'Show more'}
```
This should be translated

tomasmatus commented on src/Containers.jsx:
```diff
@@ -375,6 +376,10 @@ export let onDownloadContainerFinished = function funcOnDownloadContainerFinishe
 };
 
+export let onShowMoreEnvVariablesClick = function funcOnShowMoreEnvVariablesClick() {
+    this.setState({ showMoreEnv: !this.state.showMoreEnv });
```
Now uses `useState()` in `rederContainerEnv`

garrett commented:
Here's what it looks like my system.

![Screenshot from 2022-06-23 18-03-00](https://user-images.githubusercontent.com/10246/175344417-14c0fc87-4074-474e-816a-8292e27b0590.png)

Toolbx containers have a lot of volumes. I guess a lot of our users won't see containers like these? Or many of these? 

If you look like "angry_yonath" (leftover garbage container for cockpit-website) or "cockpit-website" (which looks the same when collapsed but is expanded in this screenshot), you can see that the variables get really wide and it shifts the content to flow to the next row.

If you compare ports and volumes, it's easy to see the internal and external ports. But it's difficult to see the internal and external volumes. Should we adjust these with the arrow like ports?

jelly commented:
I'm wondering, do we want to use the same design for ports as for volumes using an arrow to show how they are mapped.

garrett commented:
> I'm wondering, do we want to use the same design for ports as for volumes using an arrow to show how they are mapped.

Yes, that's my suggestion in the comment right before yours. :wink:

tomasmatus commented:
I see, a vertical monitor. I didn't think about it since I'm not using one.

> Should we adjust these with the arrow like ports?

> I'm wondering, do we want to use the same design for ports as for volumes using an arrow to show how they are mapped.

Sounds good!

garrett commented:
> I see, a vertical monitor. I didn't think about it since I'm not using one.

Ironically, not. I resized the page to be taller than my monitor.

It's ironic because I realized I am actually using a vertical monitor, so I don't know why I didn't just maximize the browser on my vertical screen. :facepalm:

(FWIW: Vertical monitors are great for seeing LOTS of code and quite a bit of docs. I've been using the [tiling assistant extension](https://extensions.gnome.org//extension/3733/tiling-assistant/) for GNOME, as it adds vertical splits (which can be resized just like normal splits.)

tomasmatus commented:
This layout uses `flex` prop of `FlexItem` and sets them to `flex_1`, `flex_2` and `flex_2` respectively:
![image](https://user-images.githubusercontent.com/74668142/176431147-033354a6-2c1b-48b8-9805-aedb5f5c6867.png)

Personally, I'm not a fan how spread each item is.

Alternatively I tried using `DescriptionList` with three columns and `isAutoColumnWidths`:
![image](https://user-images.githubusercontent.com/74668142/176431919-daabedb3-c190-445f-8183-3995adc12d93.png)

Here each item is closer together and aligned to the left. But maybe they are too close together.

Both of these possible layouts I experimented with fix the above mentioned issue when clicking the `Show more` button.

First layout:
![image](https://user-images.githubusercontent.com/74668142/176432942-cf78a4e3-3fef-4058-82a0-d57d59c4f6c9.png)
![image](https://user-images.githubusercontent.com/74668142/176432964-ea95a104-adce-4b84-a186-ecd8b2e63787.png)

Second layout:
![image](https://user-images.githubusercontent.com/74668142/176432512-35d191ea-3f0a-43c8-8cb6-2833e02d761c.png)
![image](https://user-images.githubusercontent.com/74668142/176432560-c4d731c5-323e-4e9b-89dc-e61c046ec894.png)

garrett commented:
Can you share either of these solutions in this PR?

If you have a toolbox container, it's a much better stress test as it does mount a lot of filesystem paths, with many of them pretty long and it also has a lot of environment variables. The standard alpine container doesn't do this, as it's intentionally very basic and is therefore not really a good test for this PR.

garrett commented:
This looks great!

However, on mobile, it's problematic:

![image](https://user-images.githubusercontent.com/10246/176679851-f408fc38-3121-473f-af3b-c247a1d458af.png)

(On the plus side, the show more/less toggle doesn't cause a shift.)

It needs some breakpoint where they're stacked vertically, which would look like this:

![image](https://user-images.githubusercontent.com/10246/176680561-b7a19170-f7dc-4f4d-9091-149d013155c9.png)

I think it needs to be something like this?

```jsx
<FlexItem flex={{ md: 'flex_1' }}>
```

garrett commented:
I pushed a change to use the commented out layout using a Description List, but only on the md breakpoint and higher, so mobile gets a stacked layout instead of side by side.

I did like the sm breakpoint better at a glance, but the directories and even environment variables can get a bit long, so I opted to use md instead.

WDYT?

(Of course, this commit should be merged in. :wink: But I wanted to keep it separate during development.)

tomasmatus commented:
I cleaned up the commits. Using `md` breakpoint seems like a good choice to me!

I also added empty state because image details and some container details are loaded slightly after. (This takes <1s on my pc so it's almost impossible to open integrations so fast but it should be there as it can make cockpit go oops).

Is it okay to use `<EmptyStatePanel />` here or should I come up with something else?

![image](https://user-images.githubusercontent.com/74668142/178452618-b7891e35-a1ab-4958-a484-bc42d8c6881e.png)

jelly commented:
> I cleaned up the commits. Using `md` breakpoint seems like a good choice to me!
> 
> I also added empty state because image details and some container details are loaded slightly after. (This takes <1s on my pc so it's almost impossible to open integrations so fast but it should be there as it can make cockpit go oops).
> 
> Is it okay to use `<EmptyStatePanel />` here or should I come up with something else?

Should be fine
 
> ![image](https://user-images.githubusercontent.com/74668142/178452618-b7891e35-a1ab-4958-a484-bc42d8c6881e.png)

Managed to also get this here, with my toolbox containers.

jelly reviewed (dismissed):
LGTM, I'll let @garrett have the final say here. Maybe the mounts should also have a show more/less if there are more then 10 but that can be an improvement. Would be nice to merge before the release next week :+1:

martinpitt commented:
Note: The test failures are unrelated, addressing them in #1042. After that lands, this PR should get a rebase, to get a clean test run.

garrett reviewed (dismissed):
> LGTM, I'll let @garrett have the final say here.

:+1: 

> Maybe the mounts should also have a show more/less if there are more then 10 but that can be an improvement. 

Nah. I don't think containers would have that many mounts. We could, of course, add it in a follow-up, but I don't think it's necessary. For example, here's a toolbox container, which uses MANY MORE mounts than an otherwise normal container:

![image](https://user-images.githubusercontent.com/10246/179755582-61a51288-821f-4fd0-be71-6054744e5ee0.png)

The reason we have show more on variables is to highlight the explicitly set variables, instead of a bunch of extra ones that are basically just noise. That's not the case for mounts in a typical container.

> Would be nice to merge before the release next week +1

Agreed! This PR has been around for a while and it's ready.

martinpitt commented:
Ah [needs updating pixel tests](https://cockpit-logs.us-east-1.linodeobjects.com/pull-1017-20220719-160602-331239af-fedora-35/log.html)

jelly commented:
```
Journal extracted to TestApplication-testCommitSystem-ubuntu-2204-127.0.0.2-2501-FAIL.log.gz
pgrep: invalid option -- 'e'

Usage:
 pgrep [options] <pattern>

Options:
 -d, --delimiter <string>  specify output delimiter
 -l, --list-name           list PID and process name
 -a, --list-full           list PID and full command line
 -v, --inverse             negates the matching
 -w, --lightweight         list all TID
 -c, --count               count of matching processes
 -f, --full                use full process name to match
 -g, --pgroup <PGID,...>   match listed process group IDs
 -G, --group <GID,...>     match real group IDs
 -i, --ignore-case         match case insensitively
 -n, --newest              select most recently started
 -o, --oldest              select least recently started
 -O, --older <seconds>     select where older than seconds
 -P, --parent <PPID,...>   match only child processes of the given parent
 -s, --session <SID,...>   match session IDs
 -t, --terminal <tty,...>  match by controlling terminal
 -u, --euid <ID,...>       match by effective IDs
 -U, --uid <ID,...>        match by real IDs
 -x, --exact               match exactly with the command name
 -F, --pidfile <file>      read PIDs from file
 -L, --logpidfile          fail if PID file is not locked
 -r, --runstates <state>   match runstates [D,S,Z,...]
 --ns <PID>                match the processes that belong to the same
                           namespace as <pid>
 --nslist <ns,...>         list which namespaces will be considered for
                           the --ns option.
                           Available namespaces: ipc, mnt, net, pid, user, uts

 -h, --help     display this help and exit
 -V, --version  output version information and exit

For more details see pgrep(1).
Connection to 127.0.0.2 closed by remote host.
```

Seems the new pgrep has issues sometimes.

jelly commented:
> Ah [needs updating pixel tests](https://cockpit-logs.us-east-1.linodeobjects.com/pull-1017-20220719-160602-331239af-fedora-35/log.html)

Pushed updated pixel tests.

martinpitt commented:
@jelly: Argh, of course `-e` is bogus for pgrep, that only makes sense for `pkill`. I'll send a PR, sorry!

martinpitt commented:
The "host path" in the "create container" dialog looks totally broken, see the new 
[reference image](https://github.com/cockpit-project/pixel-test-reference/compare/cf34d57a5726378895db9162bfad2fd17df27ca1..c0c3aa946db6c0698ada1fbeee42f5c24e8aa5c9#diff-daed45a0ac142e31a29b7d61208beaf49800e8bc835e7e09c75c0c153e55df2d):

![image](https://raw.githubusercontent.com/cockpit-project/pixel-test-reference/c0c3aa946db6c0698ada1fbeee42f5c24e8aa5c9/TestApplication-testRunImageSystem-integration-pixels.png)

This is true in all resolutions. Can we fix that before landing?

marusak commented:
> Is that from this PR? or is it like that in main? Could it be due to recent PF updates?

Not introduced in this PR, this is the case on main. Also don't think it is recent, it was the same way since we introduced pixel tests: (https://github.com/cockpit-project/cockpit-podman/pull/837)
https://github.com/cockpit-project/pixel-test-reference/blob/df7480634251052f6a93eaf6ad1a6c6b162860da/TestApplication-testRunImageSystem-integration-pixels.png

marusak reviewed (approved):
Nice, thank you!

garrett commented:
> Not introduced in this PR, this is the case on main. Also don't think it is recent, it was the same way since we introduced pixel tests: (https://github.com/cockpit-project/cockpit-podman/pull/837)

Ugh. Wow. I know this was a problem with a previous PF update and we fixed it elsewhere. I guess the fix isn't here for whatever reason. :cry: 

At least it's not a blocker for this PR, but we should address this soon.
